In [61]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import scipy.stats as stats
import statsmodels.api as sm

%matplotlib inline

5. Using the pandas read_csv() function, read the GDP dataset into your notebook as a DataFrame called gdp_df.

Take a look at the first 10 rows.

Look at the last 5 rows. Do you see a problem?

Redo the read_csv() call to correct this issue - do not modify the original csv file.

In [97]:
gdp_df = pd.read_csv("../data/gdp_per_capita.csv")

In [98]:
gdp_df.head(10)

Country or Area  Year        Value  Value Footnotes
0     Afghanistan  2018  1734.723214              NaN
1     Afghanistan  2017  1758.465636              NaN
2     Afghanistan  2016  1757.023490              NaN
3     Afghanistan  2015  1766.593077              NaN
4     Afghanistan  2014  1795.735834              NaN
5     Afghanistan  2013  1807.762344              NaN
6     Afghanistan  2012  1772.764974              NaN
7     Afghanistan  2011  1626.764793              NaN
8     Afghanistan  2010  1671.581238              NaN
9     Afghanistan  2009  1502.354073              NaN

In [99]:
gdp_df.tail(5)

Country or Area                                      Year        Value  \
6726        Zimbabwe                                      1992  2591.007534   
6727        Zimbabwe                                      1991  2906.272849   
6728        Zimbabwe                                      1990  2819.549467   
6729   footnoteSeqID                                  Footnote          NaN   
6730               2  Excludes South Sudan after July 9, 2011.          NaN   

      Value Footnotes  
6726              NaN  
6727              NaN  
6728              NaN  
6729              NaN  
6730              NaN

In [100]:
gdp_df = gdp_df[:-2] #Drop last two rows

In [101]:
gdp_df.tail()

Country or Area  Year        Value  Value Footnotes
6724        Zimbabwe  1994  2768.309953              NaN
6725        Zimbabwe  1993  2572.870395              NaN
6726        Zimbabwe  1992  2591.007534              NaN
6727        Zimbabwe  1991  2906.272849              NaN
6728        Zimbabwe  1990  2819.549467              NaN

6. Drop the 'Value Footnotes' column, and rename the remaining columns to ‘Country’, ‘Year’, and ‘GDP_Per_Capita’.

In [102]:
gdp_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6729 entries, 0 to 6728
Data columns (total 4 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Country or Area  6729 non-null   object 
 1   Year             6729 non-null   object 
 2   Value            6729 non-null   float64
 3   Value Footnotes  1 non-null      float64
dtypes: float64(2), object(2)
memory usage: 210.4+ KB


In [103]:
gdp_df.drop(columns = "Value Footnotes") # also gdp_df.drop('Value Footnotes', axis = 1)

Country or Area  Year        Value
0        Afghanistan  2018  1734.723214
1        Afghanistan  2017  1758.465636
2        Afghanistan  2016  1757.023490
3        Afghanistan  2015  1766.593077
4        Afghanistan  2014  1795.735834
...              ...   ...          ...
6724        Zimbabwe  1994  2768.309953
6725        Zimbabwe  1993  2572.870395
6726        Zimbabwe  1992  2591.007534
6727        Zimbabwe  1991  2906.272849
6728        Zimbabwe  1990  2819.549467

[6729 rows x 3 columns]

In [104]:
gdp_df.head() # changes above didn't save; assign to gdp_df

Country or Area  Year        Value  Value Footnotes
0     Afghanistan  2018  1734.723214              NaN
1     Afghanistan  2017  1758.465636              NaN
2     Afghanistan  2016  1757.023490              NaN
3     Afghanistan  2015  1766.593077              NaN
4     Afghanistan  2014  1795.735834              NaN

In [105]:
gdp_df = gdp_df.drop(columns = "Value Footnotes")

In [106]:
gdp_df.head()

Country or Area  Year        Value
0     Afghanistan  2018  1734.723214
1     Afghanistan  2017  1758.465636
2     Afghanistan  2016  1757.023490
3     Afghanistan  2015  1766.593077
4     Afghanistan  2014  1795.735834

In [107]:
gdp_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6729 entries, 0 to 6728
Data columns (total 3 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Country or Area  6729 non-null   object 
 1   Year             6729 non-null   object 
 2   Value            6729 non-null   float64
dtypes: float64(1), object(2)
memory usage: 157.8+ KB


In [108]:
gdp_df = gdp_df.rename(columns={'Country or Area': 'Country', 'Value': 'GDP_Per_Capita'})

In [109]:
gdp_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6729 entries, 0 to 6728
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Country         6729 non-null   object 
 1   Year            6729 non-null   object 
 2   GDP_Per_Capita  6729 non-null   float64
dtypes: float64(1), object(2)
memory usage: 157.8+ KB


7. How many rows and columns does gdp_df have? What are the data types of its columns? If any of the columns are not the expected types, figure out why and fix it

-gdp_df has 6729 rows and 3 columns

-the data types are "objects" and floats (numeric values)

-now I need to fix the data vaguely stored as objects

In [110]:
gdp_df["Year"] = pd.to_numeric(gdp_df["Year"])

In [111]:
gdp_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6729 entries, 0 to 6728
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Country         6729 non-null   object 
 1   Year            6729 non-null   int64  
 2   GDP_Per_Capita  6729 non-null   float64
dtypes: float64(1), int64(1), object(1)
memory usage: 157.8+ KB


In [112]:
gdp_df.shape # provides number of rows and then columns

(6729, 3)

In [114]:
gdp_df.Country.value_counts() # value counts for this column

Turkmenistan                 29
Samoa                        29
Honduras                     29
Grenada                      29
Kyrgyz Republic              29
                             ..
Nauru                        12
Sint Maarten (Dutch part)     9
South Sudan                   8
Turks and Caicos Islands      8
Djibouti                      1
Name: Country, Length: 243, dtype: int64

8: Which years are represented in this dataset? 
-The years range from 2004 to 2018

Take a look at the number of observations per year. What do you notice?
-There are missing data for some years

In [115]:
gdp_df.Year.describe()

count    6729.000000
mean     2004.293357
std         8.269834
min      1990.000000
25%      1997.000000
50%      2004.000000
75%      2011.000000
max      2018.000000
Name: Year, dtype: float64

In [116]:
gdp_df.Year.value_counts()

2011    243
2012    241
2009    241
2010    241
2013    241
2015    240
2008    240
2014    240
2007    239
2016    239
2017    239
2005    238
2004    238
2002    238
2006    238
2003    238
2001    237
2000    236
1999    228
1998    227
2018    226
1997    226
1996    224
1995    224
1994    218
1993    215
1992    213
1991    211
1990    210
Name: Year, dtype: int64

9. How many countries are represented in this dataset? 

There are 243 countries represented in the dataset

Which countries are least represented in the dataset? Why do you think these countries have so few observations?

The following 20 countries are the least represented in the dataset:
Montenegro                   22
San Marino                   21
Cayman Islands               20
Eritrea                      20
Libya                        20
Palau                        19
Timor-Leste                  19
Curaçao                     19
Liberia                      19
Qatar                        19
Kosovo                       19
Other small states           19
Small states                 19
São Tomé and Principe      18
Afghanistan                  17
Nauru                        12
Sint Maarten (Dutch part)     9
South Sudan                   8
Turks and Caicos Islands      8
Djibouti                      1



In [122]:
# of countries in dataset
unique_countries = gdp_df.drop_duplicates(subset="Country")

In [123]:
unique_countries["Country"].value_counts()

Lebanon            1
Tajikistan         1
Niger              1
Malawi             1
Sudan              1
                  ..
Solomon Islands    1
Congo              1
Sierra Leone       1
Spain              1
Moldova            1
Name: Country, Length: 243, dtype: int64

In [124]:
gdp_df.Country.value_counts().tail(20) # tail lets you see the bottom values

Montenegro                   22
San Marino                   21
Cayman Islands               20
Eritrea                      20
Libya                        20
Palau                        19
Timor-Leste                  19
Curaçao                     19
Liberia                      19
Qatar                        19
Kosovo                       19
Other small states           19
Small states                 19
São Tomé and Principe      18
Afghanistan                  17
Nauru                        12
Sint Maarten (Dutch part)     9
South Sudan                   8
Turks and Caicos Islands      8
Djibouti                      1
Name: Country, dtype: int64

10. Create a new dataframe by subsetting gdp_df to just the year 2014. Call this new dataframe gdp_2014.

In [128]:
gdp_2014 = gdp_df[gdp_df['Year'] == 2014]

In [130]:
gdp_2014

Country  Year  GDP_Per_Capita
4             Afghanistan  2014     1795.735834
21                Albania  2014    10700.993216
50                Algeria  2014    13557.769757
79                 Angola  2014     6813.052106
108   Antigua and Barbuda  2014    20440.540555
...                   ...   ...             ...
6592   West Bank and Gaza  2014     4445.307470
6617                World  2014    14514.392218
6646                Yemen  2014     3828.484887
6675               Zambia  2014     3684.685507
6704             Zimbabwe  2014     2506.490220

[240 rows x 3 columns]

11. Use .describe() to find the summary statistics for GDP per capita in 2014.

In [131]:
gdp_2014.describe()

Year  GDP_Per_Capita
count   240.0      240.000000
mean   2014.0    18034.303556
std       0.0    19320.930599
min    2014.0      672.681648
25%    2014.0     4301.187365
50%    2014.0    11868.067810
75%    2014.0    24388.408410
max    2014.0   130439.106674

12. Create a histogram of GDP Per Capita numbers for 2014 (you may wish to adjust the number of bins for your histogram). How would you describe the shape of the distribution?